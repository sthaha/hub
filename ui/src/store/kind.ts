import { types, Instance } from 'mobx-state-tree';

export const Kind = types
  .model({
    name: types.identifier,
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type IKind = Instance<typeof Kind>;
export type IKindStore = Instance<typeof KindStore>;

export const KindStore = types
  .model({
    kinds: types.map(Kind)
  })

  .actions((self) => ({
    add(item: string) {
      self.kinds.put({ name: item, selected: false });
    },

    clearSelected() {
      self.kinds.forEach((k) => {
        k.selected = false;
      });
    }
  }))

  .views((self) => ({
    get items() {
      return Array.from(self.kinds.values());
    },

    get selected() {
      const list = new Set();
      self.kinds.forEach((c: IKind) => {
        if (c.selected) {
          list.add(c.name);
        }
      });

      return list;
    }
  }));
