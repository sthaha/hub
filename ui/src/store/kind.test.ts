import { Kind, KindStore } from './kind';
import { getSnapshot } from 'mobx-state-tree';

describe('Store Object', () => {
  it('can create a kind object', () => {
    const store = Kind.create({
      name: 'Task'
    });

    expect(store.name).toBe('Task');
  });

  it('creates a kind store', (done) => {
    const store = KindStore.create({});

    const item = Kind.create({
      name: 'Task'
    });

    store.add(item.name);

    expect(store.kinds.get('Task')?.name).toBe('Task');
    expect(getSnapshot(store.kinds)).toMatchSnapshot();

    done();
  });

  it('should toggle a selected kind', (done) => {
    const store = KindStore.create({});

    const item = Kind.create({
      name: 'Task'
    });

    store.add(item.name);
    store.kinds.get('Task')?.toggle();

    expect(store.selected.size).toBe(1);
    expect(store.kinds.get('Task')?.selected).toBe(true);

    done();
  });

  it('should clear all the selected kind', (done) => {
    const store = KindStore.create({});

    const item = {
      name: 'Task'
    };

    store.add(item.name);
    store.kinds.get('Task')?.toggle();

    store.clearSelected();
    expect(store.kinds.get('Task')?.selected).toBe(false);

    done();
  });
});
