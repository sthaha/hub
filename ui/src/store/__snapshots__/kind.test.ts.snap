// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store Object creates a kind store 1`] = `
Object {
  "Task": Object {
    "name": "Task",
    "selected": false,
  },
}
`;
